### PR TITLE
Form Recognizer Preview 2 Design Proposal

### LINKED ISSUE #9699: [Design] Form Recognizer: Design Clients & Namespaces
Design clients for FormRecognizer service to support hero scenarios in a way that will work for v2.x, to the best of our knowledge, will evolve to support v3.x.

---
Prior details: 

Additional questions from Krzysztof to address as part of this:
1.	Should scenario specific clients be in one namespace or many namespaces?
2.	Should the custom client be in separate namespace from the scenario specific clients?

--

1. Multiple Clients for Custom/CustomModel/Layout/Receipt
Possible approaches, in order of FR team preference:
•	Single client with nested accessors (either properties or methods) for specialized areas
o	PRO: Clear separation of concerns
o	PRO: Reuse code and docstrings
o	CON: Deeply nested accessors
o	CON: sub-clients discouraged and/or not common in Azure SDKs
•	Multiple clients for each specialized area
o	PRO: Clear separation of concerns.
o	PRO: Reuse code and docstrings
o	CON: Duplicate XyzClientOptions for each XyzClient
o	CON: Multiple entry points for user
•	Single “flat” client with all methods available at once
o	PRO: Single entry point for user
o	PRO: Single client options class
o	CON: Verbose method names
o	CON: Duplicated method definitions and docstrings
o	CON: Duplicated method parameter for hierarchical resources (modelId)
Status: Switched from “Single client with nested accessors” to “Multiple clients for each specialized area”. Pending review.
Priority: high
Effort: high

Consider namespaces and whether to split into packages as part of this.

### LINKED ISSUE #10542: [Add methods:] Add GetCustomModel and GetCustomLabeledModel methods to CustomFormClient
These will call the endpoint: https://westus2.dev.cognitive.microsoft.com/docs/services/form-recognizer-api-v2-preview/operations/GetCustomModel

### LINKED ISSUE #10623: [Renames:] Rename raw extracted items to TextElements

### LINKED ISSUE #10643: Make source parameter to StartTraining methods a URI instead of a string
This should take a SAS URI to storage container holding training files.

Also: find a more descriptive name than `source`

Consider an overload that handles local filepaths for Docker containers.

### LINKED ISSUE #10735: Take ModelId off CustomModel
Duplicated in ModelInfo.

### PR DESCRIPTION
**Pull request by annelo-msft**
Fixes #10623
Fixes #10643
Fixes #10735
Fixes #9699
Fixes #10542

### CODE REVIEW

File: sdk/formrecognizer/Azure.AI.FormRecognizer/Azure.AI.FormRecognizer-Preview2.sln [hunk 1]
```diff
@@ -7,6 +7,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.AI.FormRecognizer", "
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.AI.FormRecognizer.Tests", "tests\Azure.AI.FormRecognizer.Tests.csproj", "{4F476D56-DDE7-43D3-8CB4-BA1E77F5A300}"
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/Azure.AI.FormRecognizer-Preview2.sln:8]: Just curious, the renaming is so that locally you could have both projects, but once decided what P1 is going to have this rename won't be necessary?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/CustomModelInfo.cs [hunk 2]
```diff
@@ -24,14 +23,14 @@ internal CustomModelInfo(ModelInfo_internal modelInfo)
 
         /// <summary>
         /// </summary>
-        public ModelStatus Status { get; }
+        public CustomFormModelStatus Status { get; }
 
         /// <summary>
         /// </summary>
-        public DateTimeOffset? CreatedOn { get; }
+        public DateTime CreatedOn { get; }
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/CustomModelInfo.cs:30]: curious, why the change to `DateTime`?

Comment (annelo-msft) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/CustomModelInfo.cs:30]: This comment from Paul in the APIView (Rev 40 on CustomModelInfo):

> bojunehsu1 week ago
> We always return UTC time (documented). So I wonder if we can just use DateTime with Kind=Utc.

We'll set the Kind to UTC.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/Azure.AI.FormRecognizer-Preview2.sln (hunk 1 above)
Comment (annelo-msft) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/Azure.AI.FormRecognizer-Preview2.sln:8]: Yes, I did this so I could have both Preview 1 and Preview 2 solutions open next to each other side by side and I wouldn't be confused about which is which.  We'll definitely put this back when it's merged back to master.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/CustomModelInfo.cs (hunk 2 above)
Comment (maririos) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/CustomModelInfo.cs:30]: Thanks for the pointer Anne.
I remembered we had a conversation about this in KeyVault and went and check and all of our client libraries use DateTimeOffset instead of DateTime. I didn’t find it in our SDK guidelines, but according to the Framework Design Guidelines:
```
 DO use DateTimeOffset whenever you are referring to an exact point in time, e.g. use it to calculate “now”, transaction times, file change times, logging event times, etc. If the time zone is not known, use it with UTC. These usages are much more common than the scenarios where DateTime is preferred, so this should be considered the default.
 DO use DateTime for any cases where the absolute point in time does not apply: e.g. store opening times that apply across time zones.
 DO use DateTime when the time zone is either not known, or is sometimes not known. This may happen in cases where it comes from a legacy data source. 
 DO NOT use DateTimeKind, if DateTimeOffset can be used instead.
DateTimeKind is an enum stored in DateTime to indicate whether the instance represents UTC, local time, or unspecified time zone. 
 DO use DateTime with a 00:00:00 time component, rather than DateTimeOffset, to represent whole dates, e.g. a date of birth.
 DO use TimeSpan to represent times of day without a date.
```
So according to that, we should go back to DateTimeOffset. What do you think?

Comment (annelo-msft) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/CustomModelInfo.cs:30]: Excellent call out - thank you for the correction, @maririos !  Would you mind posting this in the APIView and I will correct the design?

Comment (maririos) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/CustomModelInfo.cs:30]: Sure

**Comment by annelo-msft** [comment]
Closed in favor of https://github.com/Azure/azure-sdk-for-net/pull/11048
